### PR TITLE
tclPackages: init

### PR DESCRIPTION
**Pull request by fgaz**
Motivation: https://discourse.nixos.org/t/a-tcl-package-set/28071

A successor of #235614 with two key differences:

* I added a by-name hierarchy.
* I removed versioned sets. We can instead rely on the [stubs](https://wiki.tcl-lang.org/page/Stubs) mechanism (between minor versions at least). Versioned sets can be added later if we we find they are needed.

The treewide migration of existing packages is its own commit to make review easier.

To do:

* [x] Finish moving packages to tcl-modules
* [x] Documentation
* [ ] ~~Enable stubs on applicable packages.~~ Let's do this later so we avoid rebuilds in this PR.

---

Pinging people who expressed interest in this: @infinisil @nat-418 @andreabedini 
Pinging the tcl maintainer: @agbrooks 

---

> [!NOTE]
> ofborg will report some rebuilds because of the change in attrpaths, however the store hashes are the same. This pr should cause 0 actual rebuilds

### CODE REVIEW

File: pkgs/top-level/all-packages.nix
Comment (fgaz) [inline pkgs/top-level/all-packages.nix:9518]: There are two ways to migrate top-level packages that use tcl packages. This is the first. The second is to keep the regular `callPackage` and get the tcl dependency from the `tclPackages` set, like this:

```diff
--- a/pkgs/applications/version-management/fossil/default.nix
+++ b/pkgs/applications/version-management/fossil/default.nix
@@ -12,7 +12,7 @@
 , sqlite
 , ed
 , which
-, tcllib
+, tclPackages
 , withJson ? true
 }:
 
@@ -28,7 +28,7 @@ stdenv.mkDerivation (finalAttrs: {
   # required for build time tool `./tools/translate.c`
   depsBuildBuild = [ buildPackages.stdenv.cc ];
 
-  nativeBuildInputs = [ installShellFiles tcl tcllib ];
+  nativeBuildInputs = [ installShellFiles tcl tclPackages.tcllib ];
 
   buildInputs = [ zlib openssl readline which ed ]
     ++ lib.optional stdenv.isDarwin libiconv
```

Using `tclPackages.callPackage` in `all-packages.nix` is incompatible with `by-name`, so I'll go with the other option. Other maintainers seem to agree https://github.com/NixOS/nixpkgs/pull/287169#discussion_r1483153767

**Comment by fgaz** [comment]
@GrahamcOfBorg eval

**Comment by fgaz** [comment]
@GrahamcOfBorg eval

**Comment by RossComputerGuy** [comment]
Result of `nixpkgs-review pr 344716` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>31 packages built:</summary>
  <ul>
    <li>tclPackages.bwidget</li>
    <li>tclPackages.critcl</li>
    <li>tclPackages.expect</li>
    <li>tclPackages.expect.dev</li>
    <li>tclPackages.incrtcl</li>
    <li>tclPackages.incrtcl.dev</li>
    <li>tclPackages.incrtcl.man</li>
    <li>tclPackages.itktcl</li>
    <li>tclPackages.itktcl.dev</li>
    <li>tclPackages.itktcl.man</li>
    <li>tclPackages.mustache-tcl</li>
    <li>tclPackages.pdf4tcl</li>
    <li>tclPackages.tcl</li>
    <li>tclPackages.tcl-fcgi</li>
    <li>tclPackages.tcl.man</li>
    <li>tclPackages.tclPackageHook</li>
    <li>tclPackages.tclcurl</li>
    <li>tclPackages.tcllib</li>
    <li>tclPackages.tclmagick</li>
    <li>tclPackages.tclreadline</li>
    <li>tclPackages.tcltls</li>
    <li>tclPackages.tcludp</li>
    <li>tclPackages.tclvfs</li>
    <li>tclPackages.tclx</li>
    <li>tclPackages.tdom</li>
    <li>tclPackages.tix</li>
    <li>tclPackages.tk</li>
    <li>tclPackages.tk.dev</li>
    <li>tclPackages.tk.man</li>
    <li>tclPackages.tkimg</li>
    <li>tclPackages.wapp</li>
  </ul>
</details>

File: pkgs/development/tcl-modules/critcl/default.nix
Comment (RossComputerGuy) [inline pkgs/development/tcl-modules/critcl/default.nix:1]: Why is this not in the `by-name` for tcl?

**Review by RossComputerGuy — COMMENTED** [review]
This looks pretty good, I don't really have anything to complain about.

File: pkgs/development/tcl-modules/critcl/default.nix
Comment (fgaz) [reply pkgs/development/tcl-modules/critcl/default.nix:1]: Because the `callPackage` call has custom arguments, because of a dependency cycle between `critcl` and `tcllib`:

```nix
critcl = self.callPackage ../development/tcl-modules/critcl {
      tcllib = self.tcllib.override { withCritcl = false; };
    };
```

Should I move the override into `package.nix`?

Comment (RossComputerGuy) [reply pkgs/development/tcl-modules/critcl/default.nix:1]: Oh gotcha, no it's fine ig.

File: doc/languages-frameworks/tcl.section.md
Comment (SuperSandro2000) [inline doc/languages-frameworks/tcl.section.md:29]:
```suggestion
    hash = "sha256-TOouWcQc3MNyJtaAGUGbaQoaCWVe6g3BPERct/V65vk=";
```

File: doc/languages-frameworks/tcl.section.md
Comment (SuperSandro2000) [inline doc/languages-frameworks/tcl.section.md:16]: Why not default to this in mkTclDerivation and make an option to disable it for the few things that don't support it?

Comment (fgaz) [reply doc/languages-frameworks/tcl.section.md:16]: I actually plan to do this in a later PR so this one stays at 0 actual rebuilds

**Review by agbrooks — APPROVED** [review]

**Comment by fgaz** [comment]
@agbrooks thanks!

Tcl packages are being updated because of the recent 9.0 release, so conflicts are likely. Most other review requests are due to all the renames, so I will add a a changelog entry and **merge this pr on Friday** unless there are objections.

**Pull request merged**
